### DISCUSSION TITLE
add widgets in the render window for filtering

### DISCUSSION THREAD

**Discussion by laserman781** [pyvista/pyvista#1402] [Q&A] [answered]
I would like to add to be able to edit the distance q in the viewer as user input. How can I make this work? The pts_array takes only the rows that are <=q.

This is what I have written:
![image](https://user-images.githubusercontent.com/12982931/62075397-f5880b80-b212-11e9-8127-e46e44424e11.png)

**Comment by banesullivan** [comment]
Maybe try using the `pyvista.Threshold()` tool with `'shortest distance'` as the scalar:

```py
pts_array = pts[['x', 'y', 'z']].values
points = pv.PolyData(pts_array)
points['shortest distance'] = pts['shortest distance'].values

pv.Threshold(points)
```

**Comment by banesullivan** [comment]
Also, @Quadius - moving forward, would you please refrain from posting screenshots of code and instead copy/paste the code directly with [markdown formatting](https://guides.github.com/features/mastering-markdown/)?

**Comment by banesullivan** [comment]
A potential feature request form this would be to add a slider bar tool for thresholding datasets directly in the render window for both the `Plotter` and `BackgroundPlotter` classes instead od the IPython tools currently available

**Comment by laserman781** [comment]
@banesullivan 
Thank you for your reply, this helps a lot!

Now is there any way I can have multiple thresholds? I would like to be able to edit the shortest distance and the elevation clip from Issue #326? You demonstrated how I could combine the segments and the points, would it also be possible to at an elevation slider for segments, points, and include the shortest distance threshold all into one?

Thank you so much!

**Comment by banesullivan** [comment, marked as answer]
#360 will resolve this by using the slider widget with a custom callback. Here's an example:

```py
import pyvista as pv
from pyvista import examples

mesh = examples.load_uniform()

p = pv.Plotter(notebook=False)

scalars = 'Spatial Point Data'
name = 'unique_name'
p.add_mesh(mesh, name=name)
p.add_mesh(mesh.outline(),)

def func(value): 
    thresh = mesh.threshold(value, scalars=scalars)
    if thresh.n_points < 1:
        p.remove_actor(name)
    else:
        p.add_mesh(thresh, name=name)
    return

p.enable_slider_widget(callback=func, 
                       rng=mesh.get_data_range(scalars))
p.show()
```

![2019-08-15 11 16 16](https://user-images.githubusercontent.com/22067021/63112734-32ac0780-bf4e-11e9-899e-29884b45dbcc.gif)


and there is a conveinance method:

```py
p = pv.Plotter(notebook=False)
p.add_mesh_threshold(mesh)
p.add_mesh(mesh.outline())
p.show()
```

**Comment by banesullivan** [comment]
@laserman781 
> Now is there any way I can have multiple thresholds? I would like to be able to edit the shortest distance and the elevation clip from Issue #326? You demonstrated how I could combine the segments and the points, would it also be possible to at an elevation slider for segments, points, and include the shortest distance threshold all into one?

You can chain thresholds together - first threshold a mesh by one scalar then threshold that output by another scalar.

**Comment by laserman781** [comment]
This is great, thanks!
